File: leetcode/reorder-list/linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func middleNode(head *ListNode) *ListNode {
	if nil == head || nil == head.Next {
		return head
	}

	oneStep, twoStep := head, head.Next
	for ; nil != twoStep && nil != twoStep.Next; {
		twoStep = twoStep.Next.Next
		oneStep = oneStep.Next
	}

	return oneStep
}

func reverseList(head, stopAndAppend *ListNode) *ListNode {
	if nil == head || nil == head.Next {
		return head
	}

	newHead := stopAndAppend
	for ; head != stopAndAppend; {
		head.Next, newHead, head = newHead, head, head.Next
	}

	return newHead
}

func reorderList(head *ListNode)  {
	if nil == head || nil == head.Next {
		return
	}

	middle := middleNode(head)
	rightHead := reverseList(middle.Next, nil)
	middle.Next = nil

	dummyNode := &ListNode{0, nil}
	dummyNode.Next = head
	tail := dummyNode

	for ; nil != head && nil != rightHead; {
		tail.Next, tail, head = head, head, head.Next
		tail.Next, tail, rightHead = rightHead, rightHead, rightHead.Next
	}

	if nil != head {
		tail.Next = head
	}
}

File: leetcode/linked-list-cycle-ii/linked-list.cpp
#include <iostream>
#include <string>
#include <stack>
#include <queue>
#include <deque>
#include <vector>
#include <assert.h>
#include <unordered_map>

using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
  ListNode *detectCycle(ListNode *head) {
    if (NULL == head){
      return NULL;
    }

    ListNode dummyNode(0);
    dummyNode.next = head;

    ListNode *one_step = head, *two_step = head->next;
    while (NULL != two_step && one_step != two_step){
      if (head == two_step || head == two_step->next){
        return head;
      }
      if (NULL == two_step->next){
        return NULL;
      }

      two_step = two_step->next->next;
      one_step = one_step->next;
    }
    if (NULL == two_step){
      return NULL;
    }

    one_step = &dummyNode;
    while (one_step != two_step){
      one_step = one_step->next;
      two_step = two_step->next;
    }

    return one_step;
  }
};
